src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,408 @@
+  cl[1]->add_datum(j, data.row(j), true);
+}
+
+void SplitAndMergeAlgorithm::split_or_merge(const unsigned int i, const unsigned int j){
+  if(allocations[i]==allocations[j]){  
+    std::vector<unsigned int> clSplit (allocations.size()); 
+    std::shared_ptr<AbstractHierarchy> data_i(unique_values[0]->clone());
+    std::shared_ptr<AbstractHierarchy> data_j(unique_values[0]->clone());  
+    std::shared_ptr<AbstractHierarchy> data_J(unique_values[0]->clone());
+    double p_i=0; 
+    double p_j=0;
+    double p_J=0;
+    unsigned int LabI = unique_values.size();
+    
+    clSplit[i]=LabI;
+    clSplit[j]=allocations[j];
+    double q=std::exp(restricted_GS(i,j,true));
+    unsigned int z=0;
+    std::set<int> set_i=cl[0]->get_data_idx();
+    std::set<int> set_j=cl[1]->get_data_idx();
+    std::set<int>::const_iterator iter_i=set_i.cbegin();
+    std::set<int>::const_iterator iter_j=set_j.cbegin();
+    for(unsigned int k=0; k < clSplit.size(); k++){
+      if(((z<S.size())and(k==S[z]))or((k==i)or(k==j))){ //since S doesn't contain i and j but we put them in data_i or data_j
+        if(*iter_i==k){
+          if (!data_i->get_card()){ //first iteration
+            p_i+=data_i->prior_pred_lpdf(data.row(k));
+          }else{
+            p_i+=data_i->conditional_pred_lpdf(data.row(k));
+          }
+          data_i->add_datum(k, data.row(k), update_hierarchy_params());    
+          iter_i++;
+          clSplit[k]=LabI;
+        }else{
+          if (!data_j->get_card()){ //first iteration
+            p_j+=data_j->prior_pred_lpdf(data.row(k));
+          }else{
+            p_j+=data_j->conditional_pred_lpdf(data.row(k));
+          }
+          data_j->add_datum(k, data.row(k), update_hierarchy_params());
+          iter_j++;
+          clSplit[k]=allocations[j];
+        }
+        if((k!=i)and(k!=j)) z++;
+      }else{
+        clSplit[k]=allocations[k];                   
+      }
+      if(allocations[k]==allocations[j]){
+        if(!data_J->get_card()){
+          p_J+=data_J->prior_pred_lpdf(data.row(k));
+        }else{
+          p_J+=data_J->conditional_pred_lpdf(data.row(k));
+        }
+        data_J->add_datum(k, data.row(k), update_hierarchy_params());
+      }                             
+    }
+    const double p1=1/q;
+    const double alpha=mixing->get_mass_new_cluster(allocations.size(), false,
+                                       true, LabI+1); //LabI should be number of clusters - 1, alternatively unique_values.size()
+    // data_i->get_card()-1 = n. of data with LabelI different from i, data_j->get_card()-1= n. of data with LabelJ different from j (+1 because we are using gamma function)
+    const double p2=tgamma(data_j->get_card()-1-1+1)*tgamma(data_i->get_card()-1-1+1)/(S.size()+2-1)*alpha;
+    const double p3=std::exp(p_i+p_j-p_J); 
+    const double AcRa=std::min(1.0,p1*p2*p3); //acceptance ratio 
+    if(accepted_proposal(AcRa)){
+      unique_values.push_back(unique_values[0]->clone());
```

Comment:
I would move all of the code in this `if` in a separate method

Reply:
We moved all the code in the `if` in an auxiliary function, called `proposal_update_allocations`, which can be used both in the split and in the merge case. The function knows in which case it is working thanks to the boolean parameter `split`.
We decided to use a single function for both cases because the structure is exactly the same, the only things that change are the data indexes of the points that must be moved and the destination cluster.